language/compiler/ir-to-bytecode/syntax/src/lexer.rs

```diff
@@ -121,57 +123,63 @@ impl<'a> fmt::Display for Token<'a> {
 
 pub struct Lexer<'input> {
     text: &'input str,
-    consumed: usize,
-    previous_end: usize,
-    token: (usize, Token<'input>, usize),
+    prev_end: usize,
+    cur_start: usize,
+    cur_end: usize,
+    token: Tok,
 }
 
 impl<'input> Lexer<'input> {
     pub fn new(s: &'input str) -> Lexer {
         Lexer {
             text: s,
-            consumed: 0,
-            previous_end: 0,
-            token: (0, Token(Tok::EOF, ""), 0),
+            prev_end: 0,
+            cur_start: 0,
+            cur_end: 0,
+            token: Tok::EOF,
         }
     }
 
     pub fn peek(&self) -> Tok {
-        (self.token.1).0
+        self.token
     }
 
     pub fn content(&self) -> &str {
-        (self.token.1).1
+        &self.text[self.cur_start..self.cur_end]
     }
 
     pub fn start_loc(&self) -> usize {
-        self.token.0
+        self.cur_start
     }
 
     pub fn previous_end_loc(&self) -> usize {
-        self.previous_end
+        self.prev_end
     }
 
     pub fn lookahead(&self) -> Result<Tok, ParseError<usize, anyhow::Error>> {
-        let text = self.text.trim_start();
-        let whitespace = self.text.len() - text.len();
-        let start_offset = self.consumed + whitespace;
-        let (tok, _) = find_token(text, start_offset)?;
+        let text = self.text[self.cur_end..].trim_start();
+        let offset = self.text.len() - text.len();
+        let (tok, _) = find_token(text, offset)?;
         Ok(tok)
     }
 
     pub fn advance(&mut self) -> Result<(), ParseError<usize, anyhow::Error>> {
-        self.previous_end = self.token.2;
-        let text = self.text.trim_start();
-        let whitespace = self.text.len() - text.len();
-        let start_offset = self.consumed + whitespace;
-        let (tok, len) = find_token(text, start_offset)?;
-        let result = &text[..len];
-        let remaining = &text[len..];
-        let end_offset = start_offset + len;
-        self.text = remaining;
-        self.consumed = end_offset;
-        self.token = (start_offset, Token(tok, result), end_offset);
+        self.prev_end = self.cur_end;
+        let text = self.text[self.cur_end..].trim_start();
+        self.cur_start = self.text.len() - text.len();
+        let (token, len) = find_token(text, self.cur_start)?;
+        self.cur_end = self.cur_start + len;
+        self.token = token;
+        Ok(())
+    }
+
+    pub fn replace_token(
+        &mut self,
+        token: Tok,
+        len: usize,
+    ) -> Result<(), ParseError<usize, anyhow::Error>> {
+        self.token = token;
+        self.cur_end = self.cur_start + len;
```

Comment:
I think the "len" value should always be subtracted here (and there should be a comment for this function to explain that). I checked where this is called and it is passing a +1 value (and a negative value isn't even possible because of the usize type) when converting a ">>" token to ">" so advancing "cur_end" by +1 in that case seems wrong. More generally, I expect this function will only ever be used to convert a longer token to a shorter one, since the lexer's default is to choose the longest token. Subtracting an unsigned offset seems like the right thing to do.

Reply:
Ah I think you misunderstood: instead of advancing `cur_end` by 1, I'm actually setting it to `cur_start + len`. Here `len` is the length of the new token replacing the old one.
```
1   >>  3
 ~  ~~~
 |  | |
 |  | cur_end
 |  cur_start
 prev_end

// lexer.replace_token(Tok::Greater, 1)
// it becomes

1   >>  3
 ~  ~~
 |  ||
 |  |cur_end
 |  cur_start
 prev_end
```
I think ultimately the goal here is just to determine the end position of the new token. It could be done by either specifying a length directly (my current approach), or a delta between the old and new. Knowing the length of the old token these two are equivalent and I don't have a strong preference over one another.

Reply:
Yes, you're absolutely right. I did misunderstand. In my original proposal "len" was an offset but in this code, it is the length of the replacement token. Can you add a doc comment before the function to explain the meaning of the parameters?